cpp/src/io/parquet/compact_protocol_reader.cpp

```diff
@@ -388,19 +393,73 @@ class parquet_field_union_enumerator : public parquet_field {
 /**
  * @brief Functor to read a vector of structures from CompactProtocolReader
  *
+ * Uses parallel processing when there are more than 512 elements
+ *
  * @return True if field types mismatch or if the process of reading a
  * struct fails
  */
 template <typename T>
-struct parquet_field_struct_list : public parquet_field_list<T, FieldType::STRUCT> {
-  parquet_field_struct_list(int f, std::vector<T>& v)
-    : parquet_field_list<T, FieldType::STRUCT>(f, v)
+class parquet_field_struct_list : public parquet_field {
+  std::vector<T>& val;
+
+ public:
+  parquet_field_struct_list(int f, std::vector<T>& v) : parquet_field(f), val(v) {}
+
+  inline void operator()(CompactProtocolReader* cpr, int field_type)
   {
-    auto const read_value = [&val = v](uint32_t i, CompactProtocolReader* cpr) {
-      CUDF_EXPECTS(i < val.size(), "Index out of bounds");
-      cpr->read(&val[i]);
-    };
-    this->bind_read_func(read_value);
+    assert_field_type(field_type, FieldType::LIST);
+    auto const [t, n] = cpr->get_listh();
+    assert_field_type(t, FieldType::STRUCT);
+    val.resize(n);
+
+    constexpr uint32_t parallel_threshold = 512;
+    if (n >= parallel_threshold) {
+      auto const num_tasks = std::min(n, cudf::detail::host_worker_pool().get_thread_count() * 2);
+      auto const items_per_task = n / num_tasks;
+      auto const remainder      = n % num_tasks;
+
+      std::vector<std::future<void>> tasks;
+      tasks.reserve(num_tasks);
+
+      std::vector<cudf::host_span<uint8_t const>> all_ranges;
+      all_ranges.reserve(n);
+
+      std::size_t struct_idx = 0;
+      for (std::size_t task_id = 0; task_id < num_tasks; ++task_id) {
+        auto const task_size = items_per_task + (task_id < remainder ? 1 : 0);
+        auto const start_idx = struct_idx;
+        auto const end_idx   = start_idx + task_size;
+
+        if (start_idx >= n) { break; }
+
+        // Collect struct ranges for the next task
+        for (std::size_t i = start_idx; i < end_idx && i < n; ++i) {
```

Comment:
Since `end_idx < n` always at L433

```suggestion
        for (std::size_t i = start_idx; i < end_idx; ++i) {
```